File: app/src/main/java/com/galib/natorepbs2/ui/ContactListAdapter.java
package com.galib.natorepbs2.ui;

import android.view.LayoutInflater;
import android.view.ViewGroup;
import android.widget.ImageView;

import androidx.annotation.NonNull;
import androidx.databinding.BindingAdapter;
import androidx.databinding.DataBindingUtil;
import androidx.databinding.ViewDataBinding;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.galib.natorepbs2.BR;
import com.galib.natorepbs2.R;
import com.galib.natorepbs2.db.Employee;
import com.squareup.picasso.Picasso;

public class ContactListAdapter extends ListAdapter<Employee, ContactListAdapter.EmployeeViewHolder> {
    public ClickListener onClickListener;

    public interface ClickListener {
        void onClickCall(String mobile);
        void onClickEmail(String email);
    }
    public ContactListAdapter(@NonNull DiffUtil.ItemCallback<Employee> diffCallback, ClickListener listener) {
        super(diffCallback);
        onClickListener = listener;
    }

    @Override
    public int getItemViewType(int position) {
        return R.layout.item_contact;
    }

    @Override
    public EmployeeViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        LayoutInflater layoutInflater = LayoutInflater.from(parent.getContext());
        ViewDataBinding binding = DataBindingUtil.inflate(layoutInflater, viewType, parent, false);
        return new EmployeeViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull EmployeeViewHolder holder, int position) {
            holder.bind(getItem(position), onClickListener);
    }


    public static class EmployeeViewHolder extends RecyclerView.ViewHolder {
        ViewDataBinding binding;

        public EmployeeViewHolder(ViewDataBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
        @BindingAdapter("setProfilePhoto")
        public static void setProfilePhoto(ImageView imageView, String imageUrl){
            Picasso.get()
                    .load(imageUrl == null || imageUrl.length() == 0 ? null : imageUrl)
                    .placeholder(R.mipmap.ic_launcher_round)
                    .error(R.mipmap.ic_launcher_round)
                    .into(imageView);
        }
        public void bind(Employee employee, ClickListener clickListener) {
            binding.setVariable(BR.employee, employee);
            binding.setVariable(BR.clickListener, clickListener);
            binding.executePendingBindings();
        }
    }


    static class OfficerDiff extends DiffUtil.ItemCallback<Employee> {

        @Override
        public boolean areItemsTheSame(@NonNull Employee oldItem, @NonNull Employee newItem) {
            return oldItem == newItem;
        }

        @Override
        public boolean areContentsTheSame(@NonNull Employee oldItem, @NonNull Employee newItem) {
            return oldItem.getMobile().equals(newItem.getMobile());
        }
    }
}

File: app/src/main/java/com/galib/natorepbs2/ui/ContactListAdapter.kt
package com.galib.natorepbs2.ui

import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.ImageView
import androidx.databinding.BindingAdapter
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.galib.natorepbs2.BR
import com.galib.natorepbs2.R
import com.galib.natorepbs2.db.Employee
import com.galib.natorepbs2.ui.ContactListAdapter.EmployeeViewHolder
import com.squareup.picasso.Picasso

@BindingAdapter("setProfilePhoto")
fun setProfilePhoto(imageView: ImageView?, imageUrl: String?) {
    Picasso.get()
        .load(if (imageUrl == null || imageUrl.isEmpty()) null else imageUrl)
        .placeholder(R.mipmap.ic_launcher_round)
        .error(R.mipmap.ic_launcher_round)
        .into(imageView)
}

class ContactListAdapter(
    var onClickListener: ClickListener
) : ListAdapter<Employee, EmployeeViewHolder>(OfficerDiff()) {
    interface ClickListener {
        fun onClickCall(mobile: String?)
        fun onClickEmail(email: String?)
    }

    override fun getItemViewType(position: Int): Int {
        return R.layout.item_contact
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): EmployeeViewHolder {
        val layoutInflater = LayoutInflater.from(parent.context)
        val binding =
            DataBindingUtil.inflate<ViewDataBinding>(layoutInflater, viewType, parent, false)
        return EmployeeViewHolder(binding)
    }

    override fun onBindViewHolder(holder: EmployeeViewHolder, position: Int) {
        holder.bind(getItem(position), onClickListener)
    }

    class EmployeeViewHolder(var binding: ViewDataBinding) : RecyclerView.ViewHolder(
        binding.root
    ) {
        fun bind(employee: Employee?, clickListener: ClickListener?) {
            binding.setVariable(BR.employee, employee)
            binding.setVariable(BR.clickListener, clickListener)
            binding.executePendingBindings()
        }
    }

    class OfficerDiff : DiffUtil.ItemCallback<Employee>() {
        override fun areItemsTheSame(oldItem: Employee, newItem: Employee): Boolean {
            return oldItem === newItem
        }

        override fun areContentsTheSame(oldItem: Employee, newItem: Employee): Boolean {
            return oldItem.mobile == newItem.mobile
        }
    }
}